libs/evcs-service/src/main/java/uk/gov/di/ipv/core/library/client/EvcsClient.java

```diff
@@ -203,13 +213,40 @@ private void checkResponseStatusCode(HttpResponse<String> evcsHttpResponse)
     @Tracing
     private HttpResponse<String> sendHttpRequest(HttpRequest evcsHttpRequest)
             throws EvcsServiceException {
-        LOGGER.info(LogHelper.buildLogMessage("Sending HTTP request to EVCS"));
+
         try {
-            HttpResponse<String> response =
-                    httpClient.send(evcsHttpRequest, HttpResponse.BodyHandlers.ofString());
+            var response =
+                    Retry.runTaskWithBackoff(
+                            sleeper,
+                            NUMBER_OF_HTTP_REQUEST_ATTEMPTS,
+                            RETRY_DELAY_MILLIS,
+                            new RetryableTask<HttpResponse<String>>() {
+                                @Override
+                                public Optional<HttpResponse<String>> run(boolean isLastAttempt)
+                                        throws RetryException, InterruptedException {
+                                    LOGGER.info(
+                                            LogHelper.buildLogMessage(
+                                                    "Sending HTTP request to EVCS"));
+                                    try {
+                                        var res =
+                                                httpClient.send(
+                                                        evcsHttpRequest,
+                                                        HttpResponse.BodyHandlers.ofString());
+                                        var statusCode = res.statusCode();
+                                        if (!isLastAttempt
+                                                && RETRYABLE_STATUS_CODES.contains(statusCode)) {
+                                            return Optional.empty();
+                                        }
```

Comment:
This feels a bit weird - we're missing any logging etc. in the case that we retry (but I think we probably want to know if it fails the first time).

Can we move the `checkResponseStatusCode` check inside the retryable task instead?

The fact all the exceptions have to be checked makes handling them cleanly a bit of a mess, although if there's only one type per use-case it is possible to make it a generic parameter:
```
public interface RetryableTask<T, E extends Throwable> {
    /*
     * Interface for a retryable task - the implementation should return
     * an Optional.empty() to indicate that the task should be retried and
     * throw a RetryException if any errors occur
     */
   Optional<T> run(boolean isLastAttempt) throws E, InterruptedException;
}
```

I'm also wondering if there's a model where we distinguish `Retryable` from `NonRetryable` exceptions? e.g. rather than returning an `Optional` you:
- Return a value for success
- Throw `RetryableException(e)` for an exception that may be retried
- Throw `NonRetryableException(e)` for an exception that may not be retried

That way you can put the logic around when to retry inside the `Retry` helper... We might have to do some dodgy exception throwing somewhere though. It is possible to trick the compiler (e.g. https://projectlombok.org/features/SneakyThrows) if we think it'll make it much more readable.